src/GatewayAPI/Database/TransactionQuerier.cs

```diff
@@ -332,37 +344,113 @@ private async Task<List<long>> GetAccountTransactionStateVersions(AccountTransac
 
     private async Task<List<Gateway.TransactionInfo>> GetTransactions(List<long> transactionStateVersions)
     {
-        var transactionWithOperationGroups = await _dbContext.LedgerTransactions
-            .Where(t => transactionStateVersions.Contains(t.ResultantStateVersion))
-            .Include(t => t.SubstantiveOperationGroups)
-            .ThenInclude(op => op.InferredAction!.Resource)
-            .Include(t => t.SubstantiveOperationGroups) // https://stackoverflow.com/a/50898208
-            .ThenInclude(op => op.InferredAction!.Validator)
-            .Include(t => t.SubstantiveOperationGroups) // https://stackoverflow.com/a/50898208
-            .ThenInclude(op => op.InferredAction!.FromAccount)
-            .Include(t => t.SubstantiveOperationGroups) // https://stackoverflow.com/a/50898208
-            .ThenInclude(op => op.InferredAction!.ToAccount)
-            .Include(t => t.RawTransaction)
-            .OrderByDescending(lt => lt.ResultantStateVersion)
-            .AsSplitQuery() // See https://docs.microsoft.com/en-us/ef/core/querying/single-split-queries
-            .ToListAsync();
+        async Task<List<LedgerTransaction>> LoadLedgerTransactions(List<long> ids)
+        {
+            await using var localCtx = await _dbContextFactory.CreateDbContextAsync();
+
+            // we're using explicit transaction here to ensure that:
+            // a) SET LOCAL call does not leak outside of this transaction
+            // b) EF will actually reuse same database connection for both queries
+            await using var tx = await localCtx.Database.BeginTransactionAsync();
+
+            // query below tend to operate on quite significant volume of data (over 4 MiB)
+            // so in order to prevent HDD/SSD usage while applying ORDER BY clause
+            await localCtx.Database.ExecuteSqlRawAsync("SET LOCAL work_mem = '32MB';");
+
+            var result = await localCtx.LedgerTransactions
+                .Where(t => ids.Contains(t.ResultantStateVersion))
+                .Include(t => t.SubstantiveOperationGroups)
+                .OrderByDescending(lt => lt.ResultantStateVersion)
+                .ToListAsync();
+
+            await tx.CommitAsync();
+
+            return result;
+        }
+
+        async Task<Dictionary<long, Resource>> LoadResources(List<long> ids)
+        {
+            await using var localCtx = await _dbContextFactory.CreateDbContextAsync();
```

Comment:
By using separate DbContexts, I imagine we potentially use different DbConnections under the surface, and may lose the serializability assumption from querying the State Version on the same DbContext, before performing these queries. (Up till now, every Gateway API response is served from a single DbContext, which I believe is guaranteed to give read consistency).

It's fine in this case, because the tables we're querying here are static, so if we get an outdated response, we'd just miss eg a new resource, and so the query would error later on when we attempt to join back in-memory.

Something to bear in mind for Babylon though! :)

Reply:
I'll have to double check that but some notes:
1. We have to use unique DbContexts as EF simply does not allow parallel (`Task.WhenAll`) queries on a single connection.
2. EF closes (returns to pool) connection after each and every query (unless explicit transaction is in use) - I think use of separate contexts changes nothing when it comes to DbConnections.
3. I must admit I'm not sure what's the default isolation level used by our application at the moment.
4. I'm leaning more and more towards direct DbConnection usage (well... Dapper might come to the rescue) as EF clearly causes lots of confusion, produces poorly performing queries from time to time and comes with its own non-negligible set of drawbacks. This is babylon-only change suggestion! :D

Reply:
1. RE 1 - everywhere else we have just awaited each part instead of parallelising. Not ideal to be fair. 

2. Hmm yeah, I think you're right. And we use the reader endpoint which could mean each query is to a different replica, with different lag from the master - https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/Aurora.Overview.Endpoints.html#Aurora.Endpoints.Reader

3. We use the default "read committed" isolation level - https://www.postgresql.org/docs/current/transaction-iso.html - which is fine for us given our data structure and query patterns.

4. Happy with this for the Gateway API part :)

Reply:
If you're happy with 4th point (for Babylon) then I wouldn't invest much time in points 2 and 3 as they won't be that relevant when we switch to non-EF usage of single connection.